File: forge/templates/services/node-service-template/template/src/services/item.service.ts
import { AlreadyExistsError, NotFoundError } from "../lib/errors.js";
import {
	itemRepository,
	type PrismaItemRepository,
} from "../data/repositories/index.js";
import type {
	ItemCreate,
	ItemStatus,
	ItemUpdate,
	PaginatedItems,
} from "../schemas/item.schema.js";
import type { IdentityContext } from "../types/auth.js";

interface ListParams {
	identity: IdentityContext;
	skip: number;
	limit: number;
	status?: ItemStatus;
	search?: string;
}

/**
 * Service layer.
 *
 * Depends on :class:`PrismaItemRepository` only via the
 * :class:`Repository` interface — tests can substitute an in-memory
 * implementation without touching the database. The default export
 * uses the repository singleton; ``itemService.withRepository(repo)``
 * builds a service bound to a custom repo for tests.
 */

function buildService(repo: PrismaItemRepository = itemRepository) {
	return {
		async list(params: ListParams): Promise<PaginatedItems> {
			const { identity, skip, limit, status, search } = params;
			const { items, total } = await repo.list(identity, {
				skip,
				limit,
				status,
				search,
			});
			return {
				items,
				total,
				skip,
				limit,
				has_more: skip + items.length < total,
			};
		},

		async create(identity: IdentityContext, data: ItemCreate) {
			const existing = await repo.findByName(identity, data.name);
			if (existing) throw new AlreadyExistsError("Item", data.name);
			return repo.create(identity, data);
		},

		async getById(identity: IdentityContext, id: string) {
			const item = await repo.getById(identity, id);
			if (!item) throw new NotFoundError("Item", id);
			return item;
		},

		async update(identity: IdentityContext, id: string, data: ItemUpdate) {
			await this.getById(identity, id);
			if (data.name) {
				const existing = await repo.findByNameExcluding(identity, data.name, id);
				if (existing) throw new AlreadyExistsError("Item", data.name);
			}
			return repo.update(identity, id, data);
		},

		async remove(identity: IdentityContext, id: string) {
			await this.getById(identity, id);
			await repo.delete(identity, id);
		},

		withRepository(other: PrismaItemRepository) {
			return buildService(other);
		},
	};
}

const itemService = buildService();

export const list = itemService.list.bind(itemService);
export const create = itemService.create.bind(itemService);
export const getById = itemService.getById.bind(itemService);
export const update = itemService.update.bind(itemService);
export const remove = itemService.remove.bind(itemService);
export const withRepository = itemService.withRepository.bind(itemService);

File: forge/templates/services/rust-service-template/template/src/services/item_service.rs
//! Service layer.
//!
//! Depends on the [`ItemRepository`] trait — any implementation,
//! including in-memory test doubles, satisfies the bound. The default
//! wiring uses [`PgItemRepository`].

use crate::identity::IdentityContext;
use sqlx::PgPool;
use uuid::Uuid;

use crate::data::repositories::{ItemRepository, PgItemRepository};
use crate::errors::AppError;
use crate::models::{CreateItem, Item, ListParams, PaginatedResponse, UpdateItem};

pub async fn list(
    pool: &PgPool,
    identity: &IdentityContext,
    params: ListParams,
) -> Result<PaginatedResponse<Item>, AppError> {
    let repo = PgItemRepository::new(pool.clone());
    repo.list(identity, params).await
}

pub async fn create(
    pool: &PgPool,
    identity: &IdentityContext,
    data: CreateItem,
) -> Result<Item, AppError> {
    let repo = PgItemRepository::new(pool.clone());
    if repo.find_by_name(identity, &data.name).await?.is_some() {
        return Err(AppError::already_exists("Item", &data.name));
    }
    repo.create(identity, data).await
}

pub async fn get_by_id(
    pool: &PgPool,
    identity: &IdentityContext,
    id: Uuid,
) -> Result<Item, AppError> {
    let repo = PgItemRepository::new(pool.clone());
    repo.get_by_id(identity, id)
        .await?
        .ok_or_else(|| AppError::not_found("Item", id.to_string()))
}

pub async fn update(
    pool: &PgPool,
    identity: &IdentityContext,
    id: Uuid,
    data: UpdateItem,
) -> Result<Item, AppError> {
    let repo = PgItemRepository::new(pool.clone());
    repo.get_by_id(identity, id)
        .await?
        .ok_or_else(|| AppError::not_found("Item", id.to_string()))?;

    if let Some(ref name) = data.name {
        if repo
            .find_by_name_excluding(identity, name, id)
            .await?
            .is_some()
        {
            return Err(AppError::already_exists("Item", name));
        }
    }

    repo.update(identity, id, data).await
}

pub async fn delete(pool: &PgPool, identity: &IdentityContext, id: Uuid) -> Result<(), AppError> {
    let repo = PgItemRepository::new(pool.clone());
    repo.get_by_id(identity, id)
        .await?
        .ok_or_else(|| AppError::not_found("Item", id.to_string()))?;
    repo.delete(identity, id).await
}
